### ISSUE TITLE
API Key Location - how to indicate API key in maps

### ISSUE THREAD

**Issue by jerryhall** [tangrams/tangram#491]
I've searched for an hour, checked every sample I can find and have no idea where to cite my API key inside a map. I did find one instance in a tutorial but, that was for a geocoder element. Even with that key cited, I'm still getting errors indicating the 'API key is missing' so my assumption is it should be located somewhere in the map variable.  Thanks!

**Comment by jerryhall** [comment]
Alrighty then, it's an API element and I simply added as a map key:value pair as  
apiKey: 'mymapzen-key'

**Comment by meetar** [comment]
Thanks Jerry - we'll update the docs so it's nigh-impossible to miss :)

**Comment by bcamper** [comment]
Hi @jerryhall -- sorry about the confusion. The general issue here is that Tangram itself is meant to be a generic rendering engine (could be used with other tile data sources in addition to Mapzen's own) that has no direct knowledge of other Mapzen-specific components, like Mapzen Vector Tiles (which require an API key).

For common use cases, we have [Mapzen.js](https://mapzen.com/documentation/mapzen-js/), which is Mapzen's JS SDK to tie together all the individual Mapzen services. Mapzen.js takes an API key parameter that it will pass through to Tangram and other Mapzen components. 

If you want to get more hands-on with Tangram (either outside of Mapzen.js, or customize your map style), there are also many variables you can tweak within Mapzen's basemap styles: https://mapzen.com/documentation/cartography/api-reference/. [`global. sdk_api_key`](https://mapzen.com/documentation/cartography/api-reference/#sdk_api_key) is one of these.

Hope this additional detail helps.

**Comment by pnorman** [comment]
> Hi @jerryhall -- sorry about the confusion. The general issue here is that Tangram itself is meant to be a generic rendering engine (could be used with other tile data sources in addition to Mapzen's own) that has no direct knowledge of other Mapzen-specific components, like Mapzen Vector Tiles (which require an API key).

A huge 👍 for this architecture decision.

**Comment by jerryhall** [comment]
Thanks for all the feedback. I genuinely appreciate all the work you all have put into this project! I understand more about the separate architectural elements with that explanation. 

tldr; I think it may be helpful for others to learn more about the context of Tangram. 

I'm excited to have found Tangram. I've literally spent months trying to come up with a good solution with my Postgres/Geoserver/Leaflet deployment using many geographic layers (tract, zip, council districts, neighborhoods) and point data (business locations, schools, faith-recreation-health centers etc.) and the cherry on the top is crime data. And, the bottom line is there is significant drag on my GeoServer with what will be hundreds of thousands of point data. 

So, looking forward to deploying a range of maps asap.

Thanks again!

**Comment by bcamper** [comment]
Great! Please send us a link of anything you publish, always good to see what people are up to :)
